admin/handlers-post.go

```diff
@@ -1242,6 +1261,73 @@ send_response:
 	incMetric(metricAdminOK)
 }
 
+// Handler for POST request for /users/permissions
+func permissionsPOSTHandler(w http.ResponseWriter, r *http.Request) {
+	incMetric(metricAdminReq)
+	responseMessage := "OK"
+	responseCode := http.StatusOK
+	utils.DebugHTTPDump(r, settingsmgr.DebugHTTP(settings.ServiceAdmin), true)
+	vars := mux.Vars(r)
+	// Extract username and verify
+	usernameVar, ok := vars["username"]
+	if !ok || !adminUsers.Exists(usernameVar) {
+		if settingsmgr.DebugService(settings.ServiceAdmin) {
+			log.Printf("DebugService: error getting username")
+		}
+		return
+	}
+	var p PermissionsRequest
+	// Get context data
+	ctx := r.Context().Value(contextKey("session")).(contextValue)
+	// Check permissions
+	if !checkAdminLevel(ctx[ctxLevel]) {
+		responseMessage = "insuficient permissions"
+		responseCode = http.StatusForbidden
+		log.Printf("%s has %s", ctx[ctxUser], responseMessage)
+		goto send_response
+	}
+	// Parse request JSON body
+	if settingsmgr.DebugService(settings.ServiceAdmin) {
+		log.Println("DebugService: Decoding POST body")
+	}
+	if err := json.NewDecoder(r.Body).Decode(&p); err != nil {
+		responseMessage = "error parsing POST body"
+		responseCode = http.StatusInternalServerError
+		if settingsmgr.DebugService(settings.ServiceAdmin) {
+			log.Printf("DebugService: %s %v", responseMessage, err)
+		}
+	} else {
+		// Check CSRF Token
+		if checkCSRFToken(ctx[ctxCSRF], p.CSRFToken) {
+			// TODO verify environments
+			perms := users.UserPermissions{
+				Environments: p.Environments,
+				Query:        p.Query,
+				Carve:        p.Carve,
+			}
+			if err := adminUsers.ChangePermissions(usernameVar, perms); err != nil {
+				responseMessage = "error changing permissions"
+				responseCode = http.StatusInternalServerError
+				log.Printf("%s has %s", ctx[ctxUser], responseMessage)
+				goto send_response
+			}
+		} else {
+			responseMessage = "invalid CSRF token"
+			responseCode = http.StatusInternalServerError
+			if settingsmgr.DebugService(settings.ServiceAdmin) {
+				log.Printf("DebugService: %s %v", responseMessage, err)
+			}
+		}
+	}
+send_response:
```

Comment:
heh, goto.

you could make this a function that gets called everywhere you currently call goto. that'd probably be cleaner. so instead of 

```
				log.Printf("%s has %s", ctx[ctxUser], responseMessage)
				goto send_response
```

you'd have 

```
				log.Printf("%s has %s", ctx[ctxUser], responseMessage)
				sendResponse(<response args>)
                                return
```